src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -674,6 +675,76 @@ ss::future<init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
       });
 }
 
+namespace {
+
+bool is_epoch_overflow(int16_t epoch) {
+    return epoch >= std::numeric_limits<int16_t>::max();
+}
+
+bool is_valid_expected_pid(
+  const tm_transaction& tx, const model::producer_identity& expected_pid) {
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.id == expected_pid.id || ((tx.last_pid.id == expected_pid.id) && is_epoch_overflow(expected_pid.epoch));
+}
+
+bool is_valid_overflow(
+  const tm_transaction& tx, model::producer_identity expected_pid) {
+    if (!is_epoch_overflow(tx.pid.epoch)) {
+        return false;
+    }
+
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.epoch == expected_pid.epoch;
+}
+
+using current_and_last_epoch
+  = std::pair<model::producer_epoch, model::producer_epoch>;
+
+checked<current_and_last_epoch> increment_epoch(
+  const tm_transaction& tx, model::producer_identity expected_pid) {
```

Comment:
Think this implicitly assumes that the epoch won't overflow, would be nice to assert that.

Reply:
No, because we can get situation where epoch will be int16_t::max, but expected pid (from request) was prev and in this case producer retry init_producer_id request, and we should not do anything